File: backend/src/schemas/shop.py
from datetime import datetime
from decimal import Decimal
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class ShopCategory(StrEnum):
    promo_code = "promo_code"
    subscription = "subscription"
    bank_perk = "bank_perk"


class ShopItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    category: ShopCategory
    title: str
    description: str
    price: Decimal
    image_url: str | None
    brand_name: str | None
    is_recommended: bool


class PurchaseResult(BaseModel):
    purchase_id: UUID
    shop_item_id: UUID
    title: str
    price_paid: Decimal
    payload: str
    purchased_at: datetime
    new_balance: Decimal

File: frontend/src/shared/api/shop.ts
import { apiClient } from './client'

export type ShopCategory = 'promo_code' | 'subscription' | 'bank_perk'

export interface ShopItem {
  id: string
  category: ShopCategory
  title: string
  description: string
  price: string
  image_url: string | null
  brand_name: string | null
  is_recommended: boolean
}

export interface PurchaseResponse {
  purchase_id: string
  shop_item_id: string
  title: string
  price_paid: string
  payload: string
  purchased_at: string
  new_balance: string
}

export const shopApi = {
  list: (category?: ShopCategory) =>
    apiClient.get<{ items: ShopItem[] }>('/shop/items', {
      params: category ? { category } : undefined,
    }),
  purchase: (shop_item_id: string) =>
    apiClient.post<PurchaseResponse>('/shop/purchase', { shop_item_id }),
}
